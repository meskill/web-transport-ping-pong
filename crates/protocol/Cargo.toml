[package]
name = "protocol"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-trait = "0.1.68"
thiserror = "1.0.40"
tokio = "1.28.2"
tracing = "0.1.37"
wtransport = { version = "0.1.1", features = ["dangerous-configuration"] }
