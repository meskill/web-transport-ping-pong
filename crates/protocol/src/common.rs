use std::io;
use std::net::SocketAddr;

use async_trait::async_trait;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ServerError {
    #[error("IO error")]
    IoError(#[from] io::Error),
    #[error("Connection error")]
    ConnectionError,
    #[error("Communication error")]
    CommunicationError(#[from] CommunicationError),
}

#[derive(Debug, Error)]
pub enum ClientError {
    #[error("IO error")]
    IoError(#[from] io::Error),
    #[error("Connection error")]
    ConnectionError,
    #[error("Communication error")]
    CommunicationError(#[from] CommunicationError),
}

#[derive(Debug, Error)]
pub enum CommunicationError {
    #[error("Connection error")]
    ConnectionError,
    #[error("Stream error")]
    StreamError,
    #[error("Stream closed")]
    StreamClosed,
}

#[async_trait]
pub trait Server {
    async fn listen(self, address: SocketAddr) -> Result<(), ServerError>;
}

#[async_trait]
pub trait Client {
    async fn connect(self, address: SocketAddr) -> Result<(), ClientError>;
}

#[async_trait]
pub trait Communication: Send + Sync {
    async fn read(&mut self) -> Result<Vec<u8>, CommunicationError>;
    async fn write(
        &mut self,
        data: impl AsRef<[u8]> + Send + Sync,
    ) -> Result<(), CommunicationError>;
}

#[async_trait]
pub trait Handler<T: Communication>: Clone + Send + Sync + 'static {
    async fn handle(&mut self, communication: T) -> Result<(), CommunicationError>;
}
