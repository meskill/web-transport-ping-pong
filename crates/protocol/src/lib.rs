mod common;
mod web_transport;

pub use common::*;
pub use web_transport::*;
