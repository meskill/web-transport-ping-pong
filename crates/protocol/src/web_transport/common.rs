use async_trait::async_trait;
use wtransport::{
    error::{ConnectionError, StreamError},
    RecvStream, SendStream,
};

use crate::{ClientError, Communication, CommunicationError, ServerError};

pub const BUFFER_LENGTH: usize = 2048;

impl From<ConnectionError> for ServerError {
    fn from(_value: ConnectionError) -> Self {
        Self::ConnectionError
    }
}

impl From<ConnectionError> for ClientError {
    fn from(_value: ConnectionError) -> Self {
        Self::ConnectionError
    }
}

impl From<ConnectionError> for CommunicationError {
    fn from(_value: ConnectionError) -> Self {
        Self::ConnectionError
    }
}

impl From<StreamError> for CommunicationError {
    fn from(_value: StreamError) -> Self {
        Self::StreamError
    }
}

pub struct WebTransportCommunication {
    recv: RecvStream,
    send: SendStream,
    buffer: [u8; BUFFER_LENGTH],
}

impl WebTransportCommunication {
    pub fn new(recv: RecvStream, send: SendStream) -> Self {
        Self {
            recv,
            send,
            buffer: [0; BUFFER_LENGTH],
        }
    }
}

#[async_trait]
impl Communication for WebTransportCommunication {
    async fn read(&mut self) -> Result<Vec<u8>, CommunicationError> {
        let Some(bytes_read) = self.recv.read(&mut self.buffer).await? else {
          // if we haven't read anything from stream then the stream is possible closed
          return Err(CommunicationError::StreamClosed);
        };

        Ok(Vec::from(&self.buffer[..bytes_read]))
    }

    async fn write(
        &mut self,
        data: impl AsRef<[u8]> + Send + Sync,
    ) -> Result<(), CommunicationError> {
        Ok(self.send.write_all(data.as_ref()).await?)
    }
}
