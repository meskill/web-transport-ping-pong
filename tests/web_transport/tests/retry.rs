use std::{
    net::{Ipv4Addr, SocketAddr},
    time::Duration,
};

use insta::{assert_snapshot, with_settings};
use tokio::{runtime::Runtime, spawn, time::sleep};

use protocol::{Client, Server, WebTransportClient, WebTransportServer};

use crate::common::{init_tracing, SERVER_CONFIG};

mod common;

#[tokio::test]
async fn client_should_reconnect_to_server() {
    let tracing = init_tracing();
    let address = SocketAddr::new(Ipv4Addr::LOCALHOST.into(), 4433);

    let server = WebTransportServer::new(SERVER_CONFIG, server::PingPongHandler);
    let client = WebTransportClient::new("localhost", client::PingPongHandler);

    let client_handler = spawn(async move { client.connect(address).await });

    let server_runtime = Runtime::new().unwrap();

    server_runtime.spawn(async move { server.listen(address).await });

    sleep(Duration::from_millis(50)).await;
    server_runtime.shutdown_background();

    // wtransport doesn't support timeouts so we need some high value to see
    // the actual error after server has shutdown
    sleep(Duration::from_millis(11000)).await;

    let server = WebTransportServer::new(SERVER_CONFIG, server::PingPongHandler);
    let server_handler = spawn(async move { server.listen(address).await });

    sleep(Duration::from_millis(150)).await;

    client_handler.abort();
    server_handler.abort();

    with_settings!({
      filters => vec![
       (r":4433", ":[server_port]"),
       (r":\d+", ":[client_port]"),
       (r"( INFO client: Got response: \[112, 111, 110, 103\]\n)+", " INFO [communication]\n")
      ]
     }, {
      assert_snapshot!(tracing.data());
    });
}
