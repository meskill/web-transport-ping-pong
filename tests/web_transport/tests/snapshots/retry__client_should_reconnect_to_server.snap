---
source: tests/web_transport/tests/retry.rs
expression: tracing.data()
---
 INFO protocol::web_transport::server: Server started on 127.0.0.1:[server_port]
 INFO protocol::web_transport::server: New connection from: 127.0.0.1:[client_port]
 INFO client: Got response: [112, 111, 110, 103]
 INFO client: Got response: [112, 111, 110, 103]
ERROR protocol::web_transport::client: Client error: Communication error
 INFO protocol::web_transport::client: Trying to reconnect to server
 INFO protocol::web_transport::server: Server started on 127.0.0.1:[server_port]
 INFO protocol::web_transport::server: New connection from: 127.0.0.1:[client_port]
 INFO client: Got response: [112, 111, 110, 103]
 INFO client: Got response: [112, 111, 110, 103]
 INFO client: Got response: [112, 111, 110, 103]

